analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SillyMathematicalComparison.cs

```diff
@@ -0,0 +1,413 @@
+            _ = ul > verySmall; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'ulong'}}
+            _ = verySmall < ul; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'ulong'}}
+            _ = verySmall == ul; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'ulong'}}
+            _ = ul == veryBig; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'ulong'}}
+        }
+
+        public void Floats()
+        {
+            const float smallFloat = float.MinValue;
+            const double smallDouble = float.MinValue;
+
+            const float bigFloat = float.MaxValue;
+            const double bigDouble = float.MaxValue;
+
+            System.Single f = GetValue<float>();
+
+            _ = bigFloat <= f; // Compliant, not (always true) or (always false)
+            _ = bigDouble <= f; // Compliant, not (always true) or (always false)
+
+            _ = smallFloat >= f; // Compliant, not (always true) or (always false)
+            _ = smallDouble >= f; // Compliant, not (always true) or (always false)
+
+            const double veryBig = double.MaxValue;
+            const double verySmall = double.MinValue;
+
+            _ = f < veryBig; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'float'}}
+            _ = veryBig > f; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'float'}}
+            _ = f > verySmall; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'float'}}
+            _ = verySmall < f; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'float'}}
+            _ = verySmall == f; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'float'}}
+            _ = f == veryBig; // Noncompliant {{Comparison to this constant is useless; the constant is outside the range of type 'float'}}
+        }
+
+        public void ConstantInBothOperands()
+        {
+            _ = "Happy" == "Sad"; // Noncompliant {{Don't compare constant values}}
+            _ = "Ipsum" != "Lorem"; // Noncompliant {{Don't compare constant values}}
+
+            var george = true;
+            var boole = false;
+
+            _ = george == boole; // Noncompliant {{Don't compare constant values}}
+
+            var number1 = 128;
+            var number2 = 42;
+
+            _ = 42 == 256; // Noncompliant {{Don't compare constant values}}
+            _ = number1 >= 256; // Noncompliant {{Don't compare constant values}}
+            _ = 42 <= number2; // Noncompliant {{Don't compare constant values}}
+            _ = number1 != number2; // Noncompliant {{Don't compare constant values}}
```

Comment:
Please add test cases for the missing [other non-numeric const types](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#144-constants). That is enum and char.

> The type specified in a constant declaration shall be sbyte, byte, short, ushort, int, uint, long, ulong, char, float, double, decimal, bool, string, an enum_type, or a reference_type. Each constant_expression shall yield a value of the target type or of a type that can be converted to the target type by an implicit conversion ([§10.2](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#102-implicit-conversions)).

and

> Note: As described in [§11.21](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#1121-constant-expressions), a constant_expression is an expression that can be fully evaluated at compile-time. Since the only way to create a non-null value of a reference_type other than string is to apply the new operator, and since the new operator is not permitted in a constant_expression, the only possible value for constants of reference_types other than string is null. end note

Reply:
Added testcases for `enum`, `char` and a `reference type`.